File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self, root, res):
        if not root:
            return 0
        
        leftMax = self.helper(root.left, res)
        rightMax = self.helper(root.right, res)
        leftMax = max(0, leftMax)
        rightMax = max(0, rightMax)
        
        res[0] = max(res[0], root.val + leftMax + rightMax)
        
        return root.val + max(leftMax, rightMax)
        
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        res = [root.val]
        self.helper(root, res)
        ans = res[0]
        return ans
    

File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int util(TreeNode* root, int& res){
        
        if(root == NULL) return 0;
        
        int leftMax = util(root->left, res);
        int rightMax = util(root->right, res);
        leftMax = max(0, leftMax);
        rightMax = max(0, rightMax);
        
        res = max(res, root->val +leftMax + rightMax);
        
        return root->val + max(leftMax, rightMax);
    }
    
    int maxPathSum(TreeNode* root) {
        
        int res = root->val;
        util(root, res);
        return res;
    }
};